boa_engine/src/vm/opcode/iteration/iterator.rs

```diff
@@ -29,6 +29,57 @@ impl Operation for IteratorNext {
     }
 }
 
+/// `IteratorNextSetDone` implements the Opcode Operation for `Opcode::IteratorNextSetDone`
+///
+/// Operation:
+///  - Calls the `next` method of `iterator`, puts its return value on the stack
+///    and sets the `[[Done]]` value of the iterator on the call frame.
+#[derive(Debug, Clone, Copy)]
+pub(crate) struct IteratorNextSetDone;
+
+impl Operation for IteratorNextSetDone {
+    const NAME: &'static str = "IteratorNextSetDone";
+    const INSTRUCTION: &'static str = "INST - IteratorNextSetDone";
+
+    fn execute(context: &mut Context<'_>) -> JsResult<CompletionType> {
+        let next_method = context.vm.pop();
+        let iterator = context.vm.pop();
+        let mut done = true;
+        let result = next_method
+            .call(&iterator, &[], context)
+            .and_then(|next_result| {
+                next_method
+                    .as_object()
+                    .cloned()
+                    .map(IteratorResult::new)
+                    .ok_or_else(|| {
+                        JsNativeError::typ()
+                            .with_message("next value should be an object")
+                            .into()
+                    })
+                    .and_then(|iterator_result| {
+                        iterator_result.complete(context).map(|d| {
+                            done = d;
+                            context.vm.push(iterator);
+                            context.vm.push(next_method);
+                            context.vm.push(next_result);
+                            CompletionType::Normal
+                        })
+                    })
+            });
+
+        context
+            .vm
+            .frame_mut()
+            .iterators
+            .last_mut()
+            .expect("iterator on the call frame must exist")
+            .1 = done;
```

Comment:
Yes I also thought about that. I would push that until we have figured out how this works in all scenarios. This current closing of iterators works because it is limited to the destructing assignment iterators. We have to extend this to the iterators of the for-of loop itself too. When doing that we have to adjust for async iterators and control flows.
When we have solved those issues, I think we should be able to just switch over the iterator objects from the stack to the frame.